### ISSUE TITLE
RPi4: Wireless 802.11ac 5GHz performance

### ISSUE THREAD

**Issue by wagnerch** [raspberrypi/linux#4390]
**Describe the bug**
Wireless performance on the RPi4 seems to be capped at ~80Mbits/sec, this happens the same on two separate RPi4's, both in different locations, one is 5 feet from the router with an Amlogic device another foot away from it (which it can get ~250Mbits/sec).

Send vs receive, performance is the same on RPi4.  It's capped at ~80Mbits/sec.

RPi4 sitting within 5 feet of router:
```
Connecting to host 192.168.0.1, port 5201
Reverse mode, remote host 192.168.0.1 is sending
[  5] local 192.168.0.11 port 56660 connected to 192.168.0.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  8.86 MBytes  74.3 Mbits/sec
[  5]   1.00-2.00   sec  9.38 MBytes  78.7 Mbits/sec
[  5]   2.00-3.00   sec  9.38 MBytes  78.7 Mbits/sec
[  5]   3.00-4.00   sec  9.23 MBytes  77.4 Mbits/sec
[  5]   4.00-5.00   sec  9.28 MBytes  77.9 Mbits/sec
[  5]   5.00-6.00   sec  9.04 MBytes  75.9 Mbits/sec
[  5]   6.00-7.00   sec  9.11 MBytes  76.4 Mbits/sec
[  5]   7.00-8.00   sec  8.96 MBytes  75.2 Mbits/sec
[  5]   8.00-9.00   sec  9.13 MBytes  76.6 Mbits/sec
[  5]   9.00-10.00  sec  9.11 MBytes  76.4 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.24  sec  94.7 MBytes  77.6 Mbits/sec    0             sender
[  5]   0.00-10.00  sec  91.5 MBytes  76.8 Mbits/sec                  receiver


wlan0     IEEE 802.11  ESSID:...
          Mode:Managed  Frequency:5.745 GHz  Access Point: D8:47:32:74:28:52
          Bit Rate=433.3 Mb/s   Tx-Power=31 dBm
          Retry short limit:7   RTS thr:off   Fragment thr:off
          Power Management:on
          Link Quality=67/70  Signal level=-43 dBm
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0
```

Amlogic device sitting about 6~7 feet away from router, same room:
```
wlan0     IEEE 802.11  ESSID:...
          Mode:Master  Channel:155  Access Point: D8:47:32:74:28:52
          Bit Rate=390 Mb/s   Tx-Power:32 dBm
          Retry min limit:7   RTS thr:off   Fragment thr:off
          Encryption key:off
          Power Management:off
          Link Quality=5/5  Signal level=-45 dBm  Noise level=-90 dBm
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0

# iperf3 -c 192.168.0.1 -p 5201 -R
Connecting to host 192.168.0.1, port 5201
Reverse mode, remote host 192.168.0.1 is sending
[  5] local 192.168.0.41 port 47912 connected to 192.168.0.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  26.4 MBytes   222 Mbits/sec
[  5]   1.00-2.00   sec  28.4 MBytes   238 Mbits/sec
[  5]   2.00-3.00   sec  30.5 MBytes   256 Mbits/sec
[  5]   3.00-4.00   sec  28.2 MBytes   236 Mbits/sec
[  5]   4.00-5.00   sec  30.7 MBytes   258 Mbits/sec
[  5]   5.00-6.00   sec  19.8 MBytes   166 Mbits/sec
[  5]   6.00-7.00   sec  11.9 MBytes   100 Mbits/sec
[  5]   7.00-8.00   sec  9.66 MBytes  81.1 Mbits/sec
[  5]   8.00-9.00   sec  9.65 MBytes  80.9 Mbits/sec
[  5]   9.00-10.00  sec  6.78 MBytes  56.9 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.07  sec   205 MBytes   171 Mbits/sec   24             sender
[  5]   0.00-10.00  sec   202 MBytes   170 Mbits/sec                  receiver
```

**To reproduce**
Unknown, it isn't something that I particularly have noticed or cared about in the past, just ran across this anomaly troubleshooting something else.

**Expected behaviour**
I would expect the performance to be in the ball park of ~180Mbits/sec and peak out around 250Mbits/sec similar to the Amlogic device.

**Actual behaviour**
Add a clear and concise description of what actually happened.

**System**
 Copy and paste the results of the raspinfo command in to this section. Alternatively, copy and paste a pastebin link, or add answers to the following questions:

* Which model of Raspberry Pi?
Raspberry Pi 4B 8GB (also same issue on a 2GB model)

* Which OS and version (`cat /etc/rpi-issue`)?
Raspberry Pi reference 2020-05-27
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 30e2dd32ba47cc3bec15ab1413c16a17e5797775, stage4

(also same issue on LibreELEC)

* Which firmware version (`vcgencmd version`)?
8GB:
Apr 30 2021 13:45:52
Copyright (c) 2012 Broadcom
version d7f29d96450abfc77cd6cf011af1faf1e03e5e56 (clean) (release) (start)

2GB:
Apr 12 2021 15:53:21
Copyright (c) 2012 Broadcom
version 0227a0840ee8c94294cf3b2c4b2d42031a58f05a (clean) (release) (start_x)

* Which kernel version (`uname -a`)?
8GB:
Linux raspberrypi 5.10.17-v8+ #1414 SMP PREEMPT Fri Apr 30 13:23:25 BST 2021 aarch64 GNU/Linux

2GB:
Linux htpc-livingroom 5.10.42 #1 SMP Wed Jun 9 00:26:39 UTC 2021 armv7l GNU/Linux

**Logs**
Lots of this on the 8GB:
[2356123.448206] brcmfmac: CONSOLE: 2355003.361 wl0: malformed chanspec 0x0

Not a whole lot on the 2GB:
[  147.323274] hdmi-audio-codec hdmi-audio-codec.1.auto: ASoC: error at snd_soc_dai_startup on i2s-hifi: -19
[  147.323882] hdmi-audio-codec hdmi-audio-codec.1.auto: ASoC: error at snd_soc_dai_startup on i2s-hifi: -19
[47367.234062] v3d fec00000.v3d: MMU error from client CLE (4) at 0x15a21000, pte invalid

**Additional context**
Well, I have noticed that in the past "kworker/u8:2+brcmf_wq/mmc1:0001:1" will consume 25~35% of CPU, and you also see the same with iperf3 testing.  May be unrelated, but it seems to me that's pretty heavy CPU usage.  Also when under heavy network use the device will stall, ssh/shell is unresponsive for a few seconds.  But perhaps it is just congestion somewhere along the path.

**Comment by wagnerch** [comment]
If I am interpreting this right, a ~41MHz clock frequency and 4-bit wide bus would suggest a theoretical limit of 160Mbps (or is it 320Mbps if DDR?)?  That would seem to be pretty far under the 433Mbps 802.11ac spec if it is non-DDR.

I am starting to wondering if this issue is 2-fold:
1. limited because the design of the device is using SDIO based WiFi chips, is it serial (1-bit wide and not 4-bits wide)?
2. perhaps the kernel is bit-banging SDIO rather than something like DMA [Looks like DMA should be involved based on kernel source code]?  Not sure is there hardware that offloads the transfer of data or is the CPU bit-banging it [thus why I am seeing 25-35% cpu spikes on the kernel work queue]?

```
# cat /sys/kernel/debug/mmc1/ios
clock:          50000000 Hz
actual clock:   41666667 Hz
vdd:            21 (3.3 ~ 3.4 V)
bus mode:       2 (push-pull)
chip select:    0 (don't care)
power mode:     2 (on)
bus width:      2 (4 bits)
timing spec:    2 (sd high-speed)
signal voltage: 0 (3.30 V)
driver type:    0 (driver type B)
```

My original problem was dealt with by increasing the buffers in Kodi and changing the buffer mode (buffermode=1) to cache everything.  My guess is I was hitting sustained 80+Mbps spikes during high-action/high-motion scenes.

**Comment by wagnerch** [comment]
Well, apparently the theoretical limit is 160Mbps across the SDIO bus, 4-bits wide at 41.6MHz, no DDR.

https://www.raspberrypi.org/forums/viewtopic.php?f=36&t=291824&start=25#p1768600

Other users reported ~112Mbps (30% overhead) in January.  **I am hitting ~80Mbps (50% overhead), seems like a possible regression.**  Any thoughts on that front to check?

I guess if I need 802.11ac speeds then I should be looking to buy a USB adapter.  Hopefully with the Pi 5 this unfortunate situation can be addressed, especially with 802.11ax out now pushing the needle over 1Gbps.

**Comment by wagnerch** [comment]
https://www.raspberrypi.org/forums/viewtopic.php?t=294462#p1779432
